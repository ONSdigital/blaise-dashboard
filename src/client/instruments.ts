import { Instrument } from "blaise-api-node-client";
import axios from "axios";

export async function getInstruments(): Promise<Instrument[]> {
  const response = await axios.get("/api/instruments");

  return response.data;
}
