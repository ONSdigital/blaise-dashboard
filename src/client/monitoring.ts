import axios from "axios";
import { MonitoringDataModel } from "../server/monitoringDataModel";

export async function getMonitoring(): Promise<MonitoringDataModel[]> {
  const response = await axios.get("/api/monitoring");
  return response.data;
}
