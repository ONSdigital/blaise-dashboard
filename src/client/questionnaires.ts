import { Questionnaire } from "blaise-api-node-client";
import axios from "axios";

export async function getQuestionnaires(): Promise<Questionnaire[]> {
  const response = await axios.get("/api/questionnaires");

  return response.data;
}
