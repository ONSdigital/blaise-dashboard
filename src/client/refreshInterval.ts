export const refreshInterval = 10 * 1000; // 10 seconds
