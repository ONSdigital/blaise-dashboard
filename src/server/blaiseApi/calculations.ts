import {CaseStatus} from "blaise-api-node-client";
import {isCompleted, isNotCompleted} from "./outcomes";

export type CaseTotals = {
    Total: number,
    Complete: number,
    NotComplete: number
}

export type CaseCalculations = {
    CompletePercentage: number
}

export function calculateCaseTotals(cases: CaseStatus[]): CaseTotals {
    const caseTotals: CaseTotals = {
        Total: 0,
        Complete: 0,
        NotComplete: 0,
    }

    for (const eachCase of cases) {
        caseTotals.Total++
        if (isNotCompleted(eachCase.outcome)) {
            caseTotals.NotComplete++
        } else {
            caseTotals.Complete++
        }
    }
    return caseTotals
}

export function calculatePercentComplete(caseTotals: CaseTotals): number {
    if (caseTotals.Total === 0) {
        return 0.00
    }
    const percentCompleted = caseTotals.Complete / caseTotals.Total * 100
    return round2DP(percentCompleted)
}

function round2DP(numberToRound: number): number {
     return Math.round((numberToRound + Number.EPSILON) * 100) / 100
}