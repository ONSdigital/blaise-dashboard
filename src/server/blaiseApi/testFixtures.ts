import {Outcome} from "blaise-api-node-client";

export const mockQuestionnaireList = [
        {
            name: "LMS2101_CC1",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        },
        {
            name: "OPN2101A",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        },
        {
            name: "OPN2007T",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        }];

export const mockCaseList = [
            {
                primaryKey: "1000001" ,
                outcome: Outcome.Completed
            },
            {
                primaryKey: "1000002" ,
                outcome: Outcome.Completed
            },
            {
                primaryKey: "1000003" ,
                outcome: Outcome.None
            },
            {
                primaryKey: "1000004" ,
                outcome: Outcome.Partial
            },
            {
                primaryKey: "1000005" ,
                outcome: Outcome.AppointmentMade
            },
            {
                primaryKey: "1000006" ,
                outcome: Outcome.NonContact
            },
            {
                primaryKey: "1000007" ,
                outcome: 999
            }];