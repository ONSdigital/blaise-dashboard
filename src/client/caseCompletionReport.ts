import { CaseCompletionReport } from "../server/blaiseApi/caseCompletionReport";
import axios from "axios";

export async function getCaseCompletionReport(questionnaireName: string): Promise<CaseCompletionReport> {
  const response = await axios.get(`/api/reports/cases/completions/${questionnaireName}`);

  return response.data;
}
