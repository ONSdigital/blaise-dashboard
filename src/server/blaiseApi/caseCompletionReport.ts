import {calculateCaseTotals, calculatePercentComplete, CaseCalculations, CaseTotals} from "./calculations";
import {CaseStatus} from "blaise-api-node-client";


export type CaseCompletionReport = CaseTotals & CaseCalculations

export function buildCaseCompletionReport(cases: CaseStatus[]): CaseCompletionReport {
    const caseTotals = calculateCaseTotals(cases);
    const caseCompletePercentage = calculatePercentComplete(caseTotals)
    return {
        Total: caseTotals.Total,
        Complete: caseTotals.Complete,
        NotComplete: caseTotals.NotComplete,
        CompletePercentage: caseCompletePercentage
    }
}