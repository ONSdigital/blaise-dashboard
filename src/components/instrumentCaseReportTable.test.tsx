/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import InstrumentCaseReportTable from "./instrumentCaseReportTable";
import { mockInstrumentList } from "../server/blaiseApi/testFixtures";

jest.mock("../client/caseCompletionReport");
import { getCaseCompletionReport } from "../client/caseCompletionReport";
import { CaseCompletionReport } from "../server/blaiseApi/caseCompletionReport";

const getCaseCompletionReportMock = getCaseCompletionReport as jest.Mock<Promise<CaseCompletionReport>>;


describe("InstrumentCaseReportTable", () => {
    const caseCompletionReport = {
        Total: 7,
        Complete: 5,
        NotComplete: 2,
        CompletePercentage: 71.43
    };

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("renders a table row for the instrument case status", async () => {
        getCaseCompletionReportMock.mockImplementation(() => Promise.resolve(caseCompletionReport));

        render(
            <InstrumentCaseReportTable instruments={mockInstrumentList} />
        );

        expect(screen.queryByText(/Loading/i)).toBeDefined();

        await waitFor(() => {
            expect(screen.getByTestId("instrument-case-report-questionnaire-OPN2101A").textContent).toEqual("OPN2101A");
            expect(screen.getByTestId("instrument-case-report-total-OPN2101A").textContent).toEqual("7");
            expect(screen.getByTestId("instrument-case-report-complete-OPN2101A").textContent).toEqual("5 (71.43%)");
            expect(screen.getByTestId("instrument-case-report-complete-percentage-OPN2101A").innerHTML).toEqual(
                "<progress id=\"file\" value=\"71.43\" max=\"100\" role=\"progressbar\" aria-valuenow=\"71.43\" aria-valuemin=\"0\" aria-valuemax=\"100\">71.43%</progress>"
            );
            expect(screen.getByTestId("case-report-table")).toMatchSnapshot();
        });
    });

    describe("when getting case completion reports errors", () => {
        it("renders an error panel", async () => {
            getCaseCompletionReportMock.mockImplementation(() => Promise.reject("Cannot get case completion reports"));

            render(
                <InstrumentCaseReportTable instruments={mockInstrumentList} />
            );

            await waitFor(() => {
                expect(screen.getByText("Failed to get completion reports for instruments: LMS2101_CC1, OPN2101A, OPN2007T")).toBeDefined();
            });
        });
    });
});
