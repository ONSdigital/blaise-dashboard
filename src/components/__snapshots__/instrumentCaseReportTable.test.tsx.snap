// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstrumentCaseReportTable renders a table row for the instrument case status 1`] = `
<table
  class="table"
  data-testid="case-report-table"
  id="case-report-table"
>
  <thead
    class="table__head"
  >
    <tr
      class="table__row"
    >
      <th
        class="table__header"
        scope="col"
      >
        Questionnaire
      </th>
      <th
        class="table__header"
        scope="col"
      >
        Cases
      </th>
      <th
        class="table__header"
        scope="col"
      >
        Completed
      </th>
      <th
        class="table__header"
        scope="col"
      >
        Progress
      </th>
    </tr>
  </thead>
  <tbody
    class="table__body"
  >
    <tr
      class="table__row"
      data-testid="instrument-table-row"
    >
      <td
        class="table__cell"
        data-testid="instrument-case-report-questionnaire-LMS2101_CC1"
      >
        LMS2101_CC1
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-total-LMS2101_CC1"
      >
        7
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-complete-LMS2101_CC1"
      >
        5
         (
        71.43
        %)
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-complete-percentage-LMS2101_CC1"
      >
        <progress
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="71.43"
          id="file"
          max="100"
          role="progressbar"
          value="71.43"
        >
          71.43
          %
        </progress>
      </td>
    </tr>
    <tr
      class="table__row"
      data-testid="instrument-table-row"
    >
      <td
        class="table__cell"
        data-testid="instrument-case-report-questionnaire-OPN2101A"
      >
        OPN2101A
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-total-OPN2101A"
      >
        7
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-complete-OPN2101A"
      >
        5
         (
        71.43
        %)
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-complete-percentage-OPN2101A"
      >
        <progress
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="71.43"
          id="file"
          max="100"
          role="progressbar"
          value="71.43"
        >
          71.43
          %
        </progress>
      </td>
    </tr>
    <tr
      class="table__row"
      data-testid="instrument-table-row"
    >
      <td
        class="table__cell"
        data-testid="instrument-case-report-questionnaire-OPN2007T"
      >
        OPN2007T
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-total-OPN2007T"
      >
        7
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-complete-OPN2007T"
      >
        5
         (
        71.43
        %)
      </td>
      <td
        class="table__cell"
        data-testid="instrument-case-report-complete-percentage-OPN2007T"
      >
        <progress
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="71.43"
          id="file"
          max="100"
          role="progressbar"
          value="71.43"
        >
          71.43
          %
        </progress>
      </td>
    </tr>
  </tbody>
</table>
`;
